[Issue #66 + #74] Implementing "Git Reflow Refresh"

simonzhu24 opened the pull request:
Creating a new command "Git Reflow Refresh" to provide a solution to issues 73 + 66.

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,26 @@
+      current_branch = GitReflow.current_branch
+
+      `git checkout #{base_branch}`
+      `git pull --rebase #{remote_location} #{base_branch}`
```
I don't think we want the `--rebase` option here as it may require a force-push to the base branch in the future (which is not something that should ever be done on `master`).

Also, we've recently added `update_current_branch` in [GitHelpers](https://github.com/reenhanced/gitreflow/blob/master/lib/git_reflow/git_helpers.rb#L34).  Go ahead and add the extra logic to first `update_destination(base_branch)` and then perform the extra `git rebase #{base_branch}` you've added here to that method and call that from here.

As a side note, please remember to run any commands with `run` so we can avoid system calls in tests :wink:

simonzhu24 replied:
Ok, I was just following the code sample from `git reflow start`. Thanks for pointing me to `update_destination(base_branch)`, its a great idea to implement such a method in `git_reflow.rb`!

simonzhu24 commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,22 @@
+      remote_location = args.length < 2 ? "origin" : args[1]
+      base_branch = args.length < 1 ? "master" : args[0]
+      
+      GitReflow.update_destination remote_location base_branch
```
All Rspec tests are currently passing. Can replace lines 16-19 with a 1 line ternary operation, but it is a bit long and you might tell me about code readability 👍

pboling commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -40,11 +40,12 @@ def fetch_destination(destination_branch)
     end
 
-    def update_destination(destination_branch)
+    def update_destination(remote = 'origin', destination_branch)
```
If you change the signature of the method this way, you have to update all callers as well.  There is no reason to break existing code like this though since the new parameter is optional, add as the second param.  Then callers keep working as is.

simonzhu24 replied:
Sure.

pboling commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,22 @@
+long_desc <<LONGTIME
+  Performs the following:\n
+  \t$ git checkout <base_branch>\n
+  \t$ git pull --rebase <remote_location> <base_branch>\n
```
Comment needs to be fixed to reflect latest changes

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -40,11 +40,12 @@ def fetch_destination(destination_branch)
     end
 
-    def update_destination(destination_branch)
+    def update_destination(destination_branch, remote = 'origin')
```
With the rebase in here, we're doing more than updating just the destination. This not only makes the naming no longer as clear on intent, but also adds some unintended side affects to the `deliver` as it will now rebase master.

I do like the addition of the `remote` option though.  Let's add that in as a hash though so we don't have to guess as to what that value represents when we call it.  So `def update_destination(destination_branch, options = {})`.

simonzhu24 replied:
Sure. Making it "merge" as we agreed so that user does not have to use the -f flag to push.

codenamev replied:
ping on the argument `options`

codenamev replied:
@simonzhu24 ping

codenamev replied:
I would still like to see the `remote = 'origin'` argument for this method change to an `options = {}`

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -40,11 +40,12 @@ def fetch_destination(destination_branch)
-      run_command_with_label "git pull origin #{destination_branch}"
+      run_command_with_label "git pull #{remote} #{destination_branch}"
       run_command_with_label "git checkout #{origin_branch}"
+      run_command_with_label "git rebase #{destination_branch}"
```
let's move this into a new function that can encapsulate the intent better:

```ruby
def update_feature_branch(branch_name = current_branch)
# first update the base branch(options = {})
base_branch = options[:base_branch] || 'master'
feature_branch = options[:feature_branch] || current_branch
remote = options[:remote] || 'origin'
update_destination(base_branch, remote: remote)
# then update the feature branch in case there are multiple authors and has remote changes
update_current_branch
# then rebase on the base branch to bring it up to date
run_command_with_label "git rebase #{branch_name}"
end
```

simonzhu24 replied:
This seems like a bit of overkill since this method is only used in 1 place and the `update_destination` method is only called from `git_reflow.rb`. I have already abstracted out most of the logic for `base_branch`, `feature_branch`, and `remote`. 

Nonetheless, I will add this. Do we need tests for this method?

codenamev replied:
I think it makes sense to keep the rebase/merge outside the context of this method. You don't have to keep everything in my example, but was more of an outline for what I'd expect to happen in a "refresh".  It'd be fine if all this did was `update_destination` and `pull origin #{base_branch}`.  We will need tests for the new method though.

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,24 @@
+      raise "usage: git reflow refresh [base_branch_name - defaults to master] [remote_location - defaults to origin]"
+    else
+      # defaults remote_location to origin
+      remote_location = args.length < 2 ? "origin" : args[1]
```
I think the `remote` arg should come first to match that of `git pull`. I know I would mix these up when running it otherwise.

simonzhu24 replied:
Definitely, as all the Git Commands such as `git pull` and `git fetch` use this format where the remote location is passed in 1st and then the branch name 2nd.

codenamev commented:
@nhance and I were having a discussion this morning on the value of supporting this command, and whether or not it would be better served in just documentation.

Currently reflow is flexible and relatively lightweight with the expectation that the user knows what they are doing.

We can already achieve updating a feature branch with a single command `git pull origin master`.  @nhance can you voice your concerns here as well?

@simonzhu24 can you explain some common scenarios where you found the need for this feature. Maybe there are use cases we aren't thinking of.

pboling commented:
We are implementing workflows which will allow us to use a different flow strategy (interactive rebase) which will allow us to not ever have the frivolous "merge commits" you get when "updating a feature branch".  As part of that we will use this intermediate "refresh" of a feature branch if we need to get new code before we "deliver" a PR.

TL:DR we are not going to be using your git flow process, but are hoping to not have to use a total fork of the code base.  We want to make the top flexible enough for most workflows that people use.

pboling commented:
Of course this particular command can easily be added to git reflow as a plugin, so this is not crucial as part of the project at all,and wouldn't mean we need to fork.  We absolutely will need the "workflow" ability, which I am working on.

pboling commented:
I have just realized from reading this other issue https://github.com/reenhanced/gitreflow/issues/74#issue-48552299 that we won't care about the frivolous merge commits because we will be interactive rebasing (like squash merging) so the git log is still clean.

So we can merge instead of rebase, and that's fine.

codenamev commented:
We have worked on several projects where the workflows differed slightly, but enough that we weren't able to use our own tool (reflow); which is what prompted the idea.

A "refresh" of the feature branch is certainly a common practice in many workflows.  Our use of "commands" though has been to try and keep them to a minimum, and use helper methods to perform specific tasks (the abstraction for this has been a slow process).  My goal for this was to make it easier to eventually introduce the idea of "workflows" by bundling together sets of helper methods and letting users re-define the specifics of these with their own modules.  As a very crude example, [see here for a rough WIP on my first stab at it](https://github.com/reenhanced/gitreflow/commit/fe9529132d6baf0de446612c592b837bf8a3e868).  I'm not 100% sold on my own idea for how to introduce this. I'm curious to see how you're tackling it :smile:

I think what @nhance and I were mostly debating is whether or not a "refresh" needs it's own command, or if this can just live inside the review and deliver.  The footprint for this seems small enough though it may be nice to include.

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -40,11 +40,24 @@ def fetch_destination(destination_branch)
-      run_command_with_label "git checkout #{destination_branch}"
-      run_command_with_label "git pull origin #{destination_branch}"
-      run_command_with_label "git checkout #{origin_branch}"
+    def update_destination(options = {})
```
Let's leave the `destination_branch` argument how it was so we can maintain backwards compatibility.

simonzhu24 replied:
Definitely a great point!

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -40,11 +40,24 @@ def fetch_destination(destination_branch)
-      run_command_with_label "git pull origin #{destination_branch}"
-      run_command_with_label "git checkout #{origin_branch}"
+    def update_destination(options = {})
+      feature_branch = options[:feature_branch]
```
again, let's default this to `current_branch` to make sure this works as it has been

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -40,11 +40,24 @@ def fetch_destination(destination_branch)
+      run_command_with_label "git checkout #{feature_branch}"
+    end
+
+    def update_feature_branch(options = {})
```
It seems like we're doing too much here.  I don't think it makes sense to update the base branch (since we do that during delivery).  All we really need to do is make sure the feature branch is synced with remote, and then pull in any upstream changes.  That can be done in two commands: `git pull origin feature-branch && git pull remote base-branch`.  Objections?

simonzhu24 replied:
I believe you mentioned that we should update the base branch in your sample code: https://github.com/reenhanced/gitreflow/pull/164#discussion-diff-59970260

with: `update_destination(base_branch, remote: remote)` 

The point in `git reflow refresh` is to keep our local feature branch in sync with the local base branch which should be in sync with the remote base branch. `git pull origin feature-branch && git pull remote base-branch` only keeps the local feature branch in sync with the remote base + feature branches.  In order to do the behavior I described and update our local base branch, we should checkout to to the base first?

codenamev replied:
Heh, totally missed the value in syncing both local copies.  I like the idea of this more now.

nhance commented:
Sorry guys, jumping in now in support of what @codenamev has said. The idea is that we make the assumption that the user knows git and is using reflow just to automate things related to the delivery process. With an implicit "you know what you're doing", it means that we don't want to hide you from git and expect the user to know how to use git, preferring to use reflow only to collapse commands that would normally be multiple steps inside a workflow into a single step.

Fetch and update or a pull down from origin is something git does really well and we don't see the value we gain in assigning an alias to that step.

We believe keeping reflow small and reusable is good for everyone, both new coders and cranky old programmers like @codenamev and I ;)

pboling commented:
I like the suggestion for

```
git pull origin feature-branch && git pull remote base-branch
```

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -40,11 +40,24 @@ def fetch_destination(destination_branch)
+      update_destination(options)
+
+      # update feature branch in case there are multiple authors and remote changes
+      update_current_branch
```
I think we want to push up after the merge, so let's just call the pull directly so we can avoid that here `run_command_with_label "git pull origin #{current_branch}"`

simonzhu24 replied:
Sure

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -40,11 +40,24 @@ def fetch_destination(destination_branch)
+      update_current_branch
+
+      # rebase on base branch
+      run_command_with_label "git merge #{base_branch}"
```
Add in `run_command_with_label "git push origin #{current_branch}"`

simonzhu24 replied:
Sure.

codenamev commented:
Coming around to this idea.  It would be nice when coming back to older projects to just `git reflow refresh` and have it sync everything.

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,21 @@
+    # usage: git reflow refresh [remote_location - defaults to origin] [base_branch_name - defaults to master]
+    # defaults remote_location to origin
+    remote_location = args.length < 2 ? "origin" : args[1]
+    base_branch = args.length < 1 ? "master" : args[0]
```
I'm wondering if the arguments are intuitive enough.  If I came to an old project and ran `git reflow refresh last-feature`, I'd expect it to check out that feature and sync it with master.  Then again, with two arguments, I see it working as it does now.  Does that seem right to you guys?

simonzhu24 replied:
`git reflow refresh last-feature` only has 1 parameter and the 1st parameter is the `remote_location`, so this command will try to pull from the `last-feature` remote. What you ideally want in this case is `git checkout last-feature` and then `git reflow refresh origin master` in order to sync the last-feature branch with the master branch from remote origin.

We currently only `refresh` the branch that you are currently on and you would have to checkout to that branch yourself.

codenamev replied:
According to the logic in this commit, `git reflow refresh last-feature` will fetch the latest changes from `origin/last-feautre` (`base_branch` in context here) and merge them into the current working branch.  Is that the behavior you expect?

The name of this command no longer makes sense once you add arguments to the mix.  When you say `git reflow refresh other-feature`, it reads as though you want to update the `other-feature` branch.  As an example, `git pull other-feature` does exactly what you'd expect because the `pull` verb describes the action.

Since both arguments are optional (and likely to be used less often), I think it would make more sense to see them as flags.  So `git reflow refresh -r origin -b other-feature`.  This would let us automate the documentation for the command a little better as well.

simonzhu24 commented:
Hey guys, since this thread is getting a bit long, I'd like some clarification here, should we just close this diff or finish the `refresh` feature?

codenamev commented:
@simonzhu24 pinging you here.

When you're actively working on a project, chances are good that your master is already up-to-date (and even if it isn't, it's usually not more than a few commits behind and the deliver command will update it before merging your feature branch).  Also, it's common practice for us to just `git pull origin master`  before submitting a review since there could be merge conflicts we'd need to resolve.  It is often a good idea to push up your feature branch before running this so you can run both your feature and the update on CI separately in case the update from master introduces failures.  The edge-case being multiple authors, git will likely complain that your branch is behind the remote branch when you go to push to it.  The base of our concern lies in introducing a new reflow command that consists of two or less commands (in this case, it'd be better suited in a helper method).

I see value in `refresh` mostly in coming back to older projects.  Where you can say "Hey reflow, make sure my master and feature branches are up-to-date and in sync".  We often work on many different projects throughout any given month, and having a single command to ensure this would be very useful.

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,21 @@
+  \t$ git push origin <current_branch>\n
+  \t$ git merge <base_branch>\n
+LONGTIME
+arg_name '[remote_location] - remote repository name to fetch updates from (origin by default), [base_branch] - branch that you want to merge with (master by default)'
```
let's change this to two flags:
```ruby
flag [:r,:remote], default_value: "origin"
flag [:b,:branch], default_value: "master"
```

simonzhu24 replied:
Cool, will do this as a better way to do command line args!

simonzhu24 commented:
Updated and responded to comments! Please review when you have a chance.

codenamev commented on README.rdoc:
```diff
@@ -88,6 +88,16 @@ Then for your Enterprise projects, you have to setup GitReflow for each one:
+
+This command updates your feature_branch and base_branch according to the remote_location and then merges your feature_branch with the base_branch. This is just a handy command to keep your branches up to date at any point in time if someone else has committed to the base_branch or the remote.
+
+  git reflow refresh <remote_location> <base_branch_name>
```
Docs need updating here to reflect the new flag syntax.

simonzhu24 replied:
@codenamev I am actually a bit unsure as to how to use the new flag syntax. I looked at the GLI documentation online but still I am a bit unclear. I noticed that `git reflow review` also sets flags on the command, but I couldn't find examples of it being used. Could you shed some light?

codenamev replied:
I find the [GLI wiki](https://github.com/davetron5000/gli/wiki/Declaring-flags-and-switches) has better docs at times :wink:.  So with the new flags you get `git reflow refresh -r upstream -b branch-name` or `git reflow refresh --remote=upstream --branch=branch-name`

simonzhu24 replied:
Hmm. This doesn't seem to be working for me locally. Perhaps:

```
 c.flag [:r,:remote], default_value: 'origin'
 c.flag [:b,:branch], default_value: 'master'
```

is not the right way of doing it?

I keep getting:
`error: Unknown option -r`

codenamev replied:
what are you running from the command line?

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,25 @@
+desc 'refresh will merge your current branch based on the base branch'
```
No need to specify the command name in the desc as it is auto generated. Also the description is misleading. Let's go with something like "Updates and synchronizes your base branch and feature branch."

simonzhu24 replied:
ok.

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,25 @@
+  \t$ git checkout <current_branch>\n
+  \t$ git pull origin <current_branch>\n
+  \t$ git push origin <current_branch>\n
+  \t$ git merge <base_branch>\n
```
missing the final `git push <remote_location> <current_branch>` here

simonzhu24 replied:
Sure.

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,25 @@
+    # usage: git reflow refresh [remote_location - defaults to origin] [base_branch_name - defaults to master]
+    # defaults remote_location to origin
+    refresh_options = {
+      'remote' => global_options.key?[:remote] ? global_options[:remote] : 'origin',
```
since `global_options` is a hash it will just return `nil` if there is no key, so this can change to: `'remote' => global_options[:remote] || 'origin',`

simonzhu24 replied:
This is a better way of doing it!

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,25 @@
+    # defaults remote_location to origin
+    refresh_options = {
+      'remote' => global_options.key?[:remote] ? global_options[:remote] : 'origin',
+      'branch' =>  global_options.key?[:branch] ? global_options[:branch] : 'master'
```
same here

simonzhu24 replied:
Yup!

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -27,26 +27,41 @@ def get_first_commit_message
 
-    def push_current_branch
-      run_command_with_label "git push origin #{current_branch}"
+    def push_current_branch(remote = 'origin')
```
Use `options[:remote]` instead for the argument.  As a rule of thumb, if you can't tell what the argument should be without looking at the method definition, then it is better served in options.  As an example `fetch_destination` is pretty clear that the argument should be the destination branch.

simonzhu24 replied:
Sure. Great suggestions! It is a good idea to pass in the options hash when it is unclear what the parameters of the method should be.

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -27,26 +27,41 @@ def get_first_commit_message
-    def update_current_branch
-      run_command_with_label "git pull origin #{current_branch}"
-      push_current_branch
+    def update_current_branch(remote = 'origin')
```
same here with `options[:remote]`

simonzhu24 replied:
Yup

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -27,26 +27,41 @@ def get_first_commit_message
+      update_destination(base_branch, remote) 
+
+      # update feature branch in case there are multiple authors and remote changes
+      update_current_branch(remote)
```
Do you really want the `remote` option here?  Say we were updating upstream changes from a forked repository, this would try and push our current branch to the upstream remote (which often don't have access to).

simonzhu24 replied:
Actually, I don't really want to execute: `git push <remote_location> <current_branch>`. I want to leave this last step up to the user. This command just updates and refreshes his current (feature branch). When it is up-to-date, it is up to the user to push it. Then, he can specify the location that he wants to push it to.

codenamev replied:
I agree, let's yank this last run out of here.

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -27,26 +27,41 @@ def get_first_commit_message
+      run_command_with_label "git merge #{base_branch}"
+
+      # pushes feature branch
+      run_command_with_label "git push #{remote} #{current_branch}"
```
same question here

simonzhu24 replied:
Same comment as above, I am removing the last push command.

simonzhu24 commented:
@codenamev I updated the code a bit. 

Changes since last time:
1. Added function flag options -r <remote_location> -b <branch>
2. Removed the final `git push` command. `Git Refresh` refreshes the user's current branch based on their base base and remote. Pushing the branch afterwards is up to the user. This is simpler than asking the user to specify their destination location to push to, just incase the use wants to continue development on the branch.
3. Fixed `Git Reflow Review` (There was a bug where the command line flag options -t and -m where accessed through `global_options` instead of `options` and was untested). The flag options only exist in `options`.

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,27 @@
+    # usage: git reflow refresh [remote_location - defaults to origin] [base_branch_name - defaults to master]
+    # defaults remote_location to origin
+    refresh_options = {
+      :remote => options[:remote] || 'origin',
```
double check, but it seems like these options should never be `nil` thanks to the `default_value`?

simonzhu24 replied:
That is correct. You are right. Must have missed this.

codenamev commented on lib/git_reflow/commands/review.rb:
```diff
@@ -6,11 +6,11 @@
   c.flag [:m, :message], default_value: 'title'
   c.action do |global_options,options,args|
-    if global_options[:title] || global_options[:message]
+    if options[:title] || options[:message]
```
sweet! 🍰

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -27,26 +27,40 @@ def get_first_commit_message
+      update_destination(base_branch, remote) 
+
+      # update feature branch in case there are multiple authors and remote changes
+      update_current_branch(options)
```
I don't think we want to provide the same `options` here in case the user is "refreshing" a non-origin remote

simonzhu24 replied:
I don't understand? We want to provide the same `options` here as it will provide the user with the accurate `remote` and `branch` name (Both are passed in from `refresh.rb` command line flags).

codenamev replied:
Right, so let's assume the user wants to refresh from a project upstream.  The user has defined a remote `upstream`.  They're in their `new-feature` branch and they run `git reflow refresh -r upstream`.  When it gets time to update their current branch here, it will try and push their current branch to the `upstream` remote via `git push upstream new-feature`.  This is common if you're working on a fork, and typically you won't have push access to `upstream`.

simonzhu24 replied:
Exactly, good point, so I'm changing it to just:

`run_command_with_label "git pull #{remote} #{current_branch}"`

instead of running both pull and push. This way, it will just update the local branch and not push any changes out before the merge. If the user wants to push it out afterwards, its up to them!

codenamev commented:
After you address those last couple comments I'll merge away :smile:

simonzhu24 commented:
Good points in the last comment! Instead of calling `update_current_branch` which does a `git pull` and a `git push`, I just run command for `git pull`, this way the user won't attempt to push to a remote that they don't have permission to directly. 

If after the `merge`, then they can push manually themselves.

codenamev commented on lib/git_reflow/git_helpers.rb:
```diff
@@ -27,26 +27,39 @@ def get_first_commit_message
+      update_destination(base_branch, remote) 
+
+      # update feature branch in case there are multiple authors and remote changes
+      run_command_with_label "git pull #{remote} #{current_branch}"
```
I'm confused as to why we'd need to pull from the `remote` here.  Can you define the use cases?  The way I was thinking of this was that you would _always_ just pull from your default remote `origin` to update your local feature branch, but with this syntax it would update it from upstream?  Going back to the forked use-case, typically there won't be a matching feature branch on the `upstream` remote repo.

simonzhu24 replied:
It is up to the user what they call their remote branches. In `git remote -v`, they can call / name their remote branches anything they want. Although it is convention to call the remote branches `origin,` that is not a requirement. I think we should offer this flexibility just incase the users don't adhere to convention.

codenamev replied:
I think that's a convention we want to adhere to.  If you did a `grep -r origin ./lib` you'll see that we heavily rely on that convention.  Also, I think this is an edge case we don't want to support as it's not a typical workflow.  We really like this feature because of it's ability to update "upstream" repos easily.  With this in place, we lose that.

simonzhu24 replied:
Thanks for the comments. I see your point now, when updating (pulling) the feature_branch, we want to stick to standard `origin`. I'm making the change right now!

simonzhu24 commented:
Ready for merge! ^_^

simonzhu24 commented:
![yes](https://cloud.githubusercontent.com/assets/18274144/14690779/a478c662-0702-11e6-8be3-a0228f7d8fa1.gif)

codenamev commented on lib/git_reflow/commands/refresh.rb:
```diff
@@ -0,0 +1,28 @@
+  \t$ git checkout <base_branch>\n
+  \t$ git pull <remote_location> <base_branch>\n
+  \t$ git checkout <current_branch>\n
+  \t$ git pull <remote_location> <current_branch>\n
```
need to update the docs here to replace `remote_location` with `origin`

simonzhu24 replied:
Done!!!

codenamev commented:
just that small wording update, other than that this LGTM!  Great work here @simonzhu24 😄

simonzhu24 commented:
![image](https://cloud.githubusercontent.com/assets/18274144/14718589/7503f19a-07ac-11e6-957d-0ef5ffd17ef4.png)
